### ISSUE TITLE
[BUG] 3.11 fails to start with segmentation fault on FreeBSD

### ISSUE THREAD

**Issue by yasuhirokimura** [rspamd/rspamd#5258]
While trying to update FreeBSD Rspamd port (mail/rspamd) to 3.11, I noticed 3.11 fails to start with segmentation fault as follwing.

```
# service rspamd start
Starting rspamd.
2024-12-17 06:21:59 #77528(main) <9dd190>; main; main: rspamd 3.11.0 is loading configuration, build id: release
Segmentation fault (core dumped)
/usr/local/etc/rc.d/rspamd: WARNING: failed to start rspamd
#
```

It can be reproduced with following steps.

1. Make clean install of FreeBSD 14.2-RELEASE amd64
2. Login as root
3. Install git package with `pkg install git`
4. `cd /usr/ports`
5. Checkout latest main branch of ports tree with `git clone https://git.freebsd.org/ports.git .`
6. Apply patch attached to https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=283365
7. Uninstall all packages with `pkg-static delete -af`
8. `cd /usr/ports/mail/rspamd`
9. `make install`
10. Add `rspamd_enable="YES"` to `/etc/rc.conf`
11. Start Rspamd with `service rspamd start`

**Comment by fraenki** [comment]
Probably fixed by https://github.com/rspamd/rspamd/commit/ccb45df90df60fae36b9438cfb2b0088e590306b.

**Comment by nlegaillart** [comment]
> Probably fixed by [ccb45df](https://github.com/rspamd/rspamd/commit/ccb45df90df60fae36b9438cfb2b0088e590306b).

Unfortunately, this patch does not work for me. I just installed the final version of the port, released today, which include it, and I get the same error message as above (I'm also running FreeBSD 14.2-RELEASE amd64)

**Comment by vstakhov** [comment]
Unfortunately, we can't provide any more help without a proper core file/backtrace. Closing issue.

**Comment by nlegaillart** [comment]
Oh, it's a real pleasure trying to troubleshoot bugs with you :( I'm not able to generate a core file right now, if I can do it I will post it here (if you do want it, of course...)

As a side note, it may be worth noting that downgrading to 10.0.2 still end to a segmentation fault at start, so it may be related to something else.

**Comment by atroxix** [comment]
I've been getting segfault since release 3.9, so I'm stuck on 3.8.4.

Here's a core dump file - is this OK?
[rspamd-3.11.0.core.gz](https://github.com/user-attachments/files/18233067/rspamd-3.11.0.core.gz)

**Comment by vstakhov** [comment]
No, I can't do anything with your core file: https://rspamd.com/doc/faq.html#what-can-i-do-with-core-files
In fact, you need to build your port with debug, like `WITH_DEBUG=1 DEBUG=1 STRIP=` in your make arguments/environment.

**Comment by atroxix** [comment]
> In fact, you need to build your port with debug, like `WITH_DEBUG=1 DEBUG=1 STRIP=` in your make arguments/environment.

I tried to do this, but didn't succeed. I also installed rspamd-devel port as it has ```-DDEBUG=ON``` in its Makefile, but this didn't result in any more useful core dump file either.

If you know how to enable DEBUG, could you please provide those instructions?

**Comment by Oclair** [comment]
> No, I can't do anything with your core file: https://rspamd.com/doc/faq.html#what-can-i-do-with-core-files In fact, you need to build your port with debug, like `WITH_DEBUG=1 DEBUG=1 STRIP=` in your make arguments/environment.

Aaah so that's why it didn't get fixed....

building rspamd with Debug is the workaround rofl......

PLEASE REOPEN THIS ISSUE!

**Comment by Oclair** [comment]
attempting to apply the workaround again...  
#cd /usr/ports/mail/rspamd && make WITH_DEBUG=yes DEBUG=yes STRIP=
#make reinstall clean

Seems workaround still works!
[12:53:19]  <Oclair> 2025-03-06 12:53:04 #33638(main) <5ed9d4>; main; main: rspamd 3.11.0 is loading configuration, build id: release

**Comment by vstakhov** [comment]
> Unfortunately, we can't provide any more help without a proper core file/backtrace. Closing issue.

Do we have anything new on that? If no, I will not reopen.

**Comment by Oclair** [comment]
Ok I will suggest that the FreeBSD port adds debug as a default build option #facepalm

**Comment by vstakhov** [comment]
@yasuhirokimura In fact, I think we should just use system libfmt to avoid this issue. Rspamd will work with libfmt 10 without any issues, and Gentoo uses that as well.

**Comment by atroxix** [comment]
> [@yasuhirokimura](https://github.com/yasuhirokimura) In fact, I think we should just use system libfmt to avoid this issue. Rspamd will work with libfmt 10 without any issues, and Gentoo uses that as well.

Does it mean that rspamd should work if libfmt-10 is installed? It doesn't in my system.

**Comment by vstakhov** [comment]
No, it should work if you have **NO** libfmt installed anywhere during the port build process (e.g. in an isolated environment provided by poudriere). If we decide to use system wide libfmt, it will also work, but a clash of both causes issues.

**Comment by vstakhov** [comment]
In fact, let's reopen this issue as FreeBSD port is due to be updated anyway.

**Comment by atroxix** [comment]
> No, it should work if you have **NO** libfmt installed anywhere during the port build process (e.g. in an isolated environment provided by poudriere). If we decide to use system wide libfmt, it will also work, but a clash of both causes issues.

Indeed. I just deleted libfmt port because nothing else required it, recompiled rspamd and now it works!

**Comment by moisseev** [comment]
What if we declare `libfmt` as a build conflict in the Rspamd port’s Makefile as a temporary measure until a more permanent fix can be applied?

**Comment by vstakhov** [comment]
I think dependency is a more preferred option here, as Gentoo uses it without any problems.

**Comment by Oclair** [comment]
I am confused.   

There is libfmt currently installed on my system (FreeBSD 13.4-RELEASE-p4) yet some ports I built on my system require it for example mariadb1011-server  https://www.freshports.org/databases/mariadb1011-server/

I currently have to rebuild rspamd with debug flags manually so it can run and NOT coredump.

**Comment by arkamar** [comment]
Yes, we use system `libfmt` in Gentoo, see https://github.com/gentoo/gentoo/blob/4da7f02e9988997ed61ee29aaab443390119299d/mail-filter/rspamd/rspamd-3.11.1.ebuild#L101, without any known issue. It is worth to say that MariaDB is quite picky about the `libfmt` version and vanilla Rspamd requires libfmt 11, therefore, we patch Rspamd for older libfmt versions compatibility, see https://github.com/gentoo/gentoo/blob/4da7f02e9988997ed61ee29aaab443390119299d/mail-filter/rspamd/files/rspamd-3.9.0-older-libfmt.patch.

**Comment by vstakhov** [comment]
Ah, that might be the issue. FreeBSD ports also have libfmt 10 and I will not patch Rspamd for that compatibility as it's utterly stupid.

**Comment by vstakhov** [comment]
In fact, I'm inclined to remove any possibility to use the system libfmt and nail the embedded version. It will get rid of any potential issues in the future at no price. Is there **any** benefit in using of the system-wide libfmt counting that it's a header only library? I can hardly find any...

**Comment by arkamar** [comment]
If there is any problem with libfmt source from downstream point of view (security issue, compatibility issue (we support multiple libcs), incompatibility with toolchain (e.g. modern GCC or Clang errors due to hardening)) we can solve it only in the system instance, but we don't have to do it multiple times for every project which bundles it.

As for Gentoo, we prefer system libfmt.

Btw, I tried to build rspamd with `-DSYSTEM_FMT=OFF` while having libfmt-10.2.1 installed and it does not segfault for me.

**Comment by vstakhov** [comment]
> As for Gentoo, we prefer system libfmt.

Since you already patch Rspamd to support ancient libfmt, it won't be a big deal if you patch it one more time, iiuc.

> If there is any problem with libfmt source from downstream point of view (security issue, compatibility issue (we support multiple libcs), incompatibility with toolchain (e.g. modern GCC or Clang errors due to hardening)) we can solve it only in the system instance, but we don't have to do it multiple times for every project which bundles it.

For now, I see the opposite issue. It's super easy for me to maintain bundled dependencies but it is just the pain in the ass to support all the bloody zoo of the versions that can be installed in different systems. So I'm not with you here.

**Comment by arkamar** [comment]
> > As for Gentoo, we prefer system libfmt.
> 
> Since you already patch Rspamd to support ancient libfmt, it won't be a big deal if you patch it one more time, iiuc.

Well, it's pita to have local patches and my plan was to drop the current one when libfmt-11 will be stable in Gentoo (it was stabilized quite recently).

> > If there is any problem with libfmt source from downstream point of view (security issue, compatibility issue (we support multiple libcs), incompatibility with toolchain (e.g. modern GCC or Clang errors due to hardening)) we can solve it only in the system instance, but we don't have to do it multiple times for every project which bundles it.
> 
> For now, I see the opposite issue. It's super easy for me to maintain bundled dependencies but it is just the pain in the ass to support all the bloody zoo of the versions that can be installed in different systems. So I'm not with you here.

I understand your point of view as a developer :)

**Comment by vstakhov** [comment]
In fact, if we are talking about header only dependencies, I think bundling is better from any point. We don't have any benefits from using a system headers - they will just be embedded like bundled one (so no code sharing). We won't even get better compatibility - if you have some issues with gcc/hardening/etc, you can just open issue upstream and I will just update bundled dependency once the upstream has fixed their issues. Hence, I see **completely** zero benefits from using the system-wide header only dependency, but I can clearly see the disadvantages on the contrary.
